sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveStrategies.scala

```diff
@@ -237,21 +237,75 @@ private[hive] trait HiveStrategies {
+        case (true, true) => null
+        case (true, false) => right
+        case (false, true) => left
+        case (false, false) =>
+          if(op_type == "or")
+            Or(left, right)
+          else if (op_type == "and")
+            And(left, right)
+          else
+            null
+      }
+    }
+
+    def resolveAndExpression(expr: Expression, partitionKeyIds: AttributeSet): Expression = {
+      expr match {
+        case and: And =>
+          constructBinaryOperators(resolveAndExpression(and.left, partitionKeyIds), resolveAndExpression(and.right, partitionKeyIds), "and")
+        case _ =>
+          resolvePredicatesExpression(expr, partitionKeyIds)
+      }
+    }
+
+    def resolveOrExpression(or: Or, partitionKeyIds: AttributeSet): Expression = {
+      (or.left.isInstanceOf[Or],or.right.isInstanceOf[Or]) match {
+        case (true, true) => constructBinaryOperators(resolveOrExpression(or.left.asInstanceOf[Or], partitionKeyIds) , resolveOrExpression(or.right.asInstanceOf[Or], partitionKeyIds), "or")
+        case (true, false) => constructBinaryOperators(resolveOrExpression(or.left.asInstanceOf[Or], partitionKeyIds) , resolveAndExpression(or.right, partitionKeyIds), "or")
+        case (false, true) => constructBinaryOperators(resolveAndExpression(or.left, partitionKeyIds) , resolveOrExpression(or.right.asInstanceOf[Or], partitionKeyIds), "or")
+        case (false, false) => constructBinaryOperators(resolveAndExpression(or.left, partitionKeyIds) , resolveAndExpression(or.right, partitionKeyIds), "or")
+      }
+    }
+
+    def resolvePredicatesExpression(expr: Expression, partitionKeyIds: AttributeSet): Expression ={
+      if(!expr.references.isEmpty && expr.references.subsetOf(partitionKeyIds))
+        expr
+      else
+        null
+    }
+
+    def extractPushDownPredicate(predicates: Seq[Expression], partitionKeyIds: AttributeSet): Seq[Expression] ={
+      predicates.map {
+        case or: Or =>
+          resolveOrExpression(or, partitionKeyIds)
+        case predicate =>
+          resolvePredicatesExpression(predicate, partitionKeyIds)
+      }
+    }
+
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
       case PhysicalOperation(projectList, predicates, relation: HiveTableRelation) =>
         // Filter out all predicates that only deal with partition keys, these are given to the
         // hive table scan operator to be used for partition pruning.
         val partitionKeyIds = AttributeSet(relation.partitionCols)
-        val (pruningPredicates, otherPredicates) = predicates.partition { predicate =>
+        val (_, otherPredicates) = predicates.partition { predicate => {
           !predicate.references.isEmpty &&
-          predicate.references.subsetOf(partitionKeyIds)
+            predicate.references.subsetOf(partitionKeyIds)
         }
+        }
+
+        val extractedPruningPredicates = extractPushDownPredicate(predicates, partitionKeyIds)
+          .filter(_ != null)
```

Comment:
@AngersZhuuuu, just for clarification, this code path does support OR expression but you want to do a partial pushdown right? Considering it needs a lot of codes as @wangyum pointed out, I think we should better try to promote to use (or convert) Spark's Parquet or ORC. It looks like an overkill to me.

Reply:
@HyukjinKwon  What I do is to extract condition's about partition keys.For the old code :
` val (pruningPredicates, otherPredicates) = predicates.partition { predicate =>
          !predicate.references.isEmpty &&
          predicate.references.subsetOf(partitionKeyIds) }`

If in expression, there contains other key, it won't be a push to HiveTableScanExec, So what I to it to fix this situation, just extract all condition about partition keys, then push it to HiveTableScanExec, HiveTableScanExec will handle complex combine expressions.

Reply:
@HyukjinKwon 
Spark's Parquet or ORC is perfect,  and it can push down filter condition, but it can't resolve the problem  that when we read a Hive table,  our first behavior is scan, What this pr want to do is to reduce the time of  resolve file info and partition metadata, and the file we scan.  Then the file num or partition num is big, it takes too long.

Reply:
I think we convert Hive table reading operations into Spark's ones, via, for instance, `spark.sql.hive.convertMetastoreParquet` conf. If the diff is small, I might be fine but this does look like an overkill to me. I haven't taken a close look but it virtually looks like we need a fix like https://github.com/apache/spark/pull/24598

I won't object if some other committers are fine with that.

Reply:
@HyukjinKwon  I know that  it's better to us convert hive metastore to DataSource API, but it can't fix all situation. In our production env, we just  change hive data's default storage type  to orc.  For partition table, if different partition;s serde is not the same, Convert will failed thing using convert, it will check all partition by table level serde.